### PR TITLE
Add election code and examples

### PR DESCRIPTION
**Pull request by brendandburns**

### CODE REVIEW

**Comment by mikedanese** [comment]
Can we use the leader election client?

https://godoc.org/k8s.io/kubernetes/pkg/client/leaderelection

**Comment by brendandburns** [comment]
oh, sure, I missed that you'd added that...  I'll refactor this in terms of that.

**Comment by brendandburns** [comment]
@mikedanese 
Switched to use the leaderelection stuff.  Had to wrap it up a bit to get a cleaner API, we should migrate that code back into the leaderelection package.

ptal,
thanks!
--brendan

File: election/example/main.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+	fn := func(str string) {
+		leader.Name = str
+		fmt.Printf("%s is the leader\n", leader.Name)
+	}
+
+	e, err := election.NewElection(*name, *id, *namespace, *ttl, fn, kubeClient)
+	if err != nil {
+		glog.Fatalf("failed to create election: %v", err)
+	}
+	go e.Run()
```
Comment (mikedanese) [inline election/example/main.go]: the e.Run() should be warpped in an Until Forever. A single election client does one cycle of candiate->aquire master->lose master and is expected to compelete.

File: election/lib/election.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+		OnStartedLeading: func(stop <-chan struct{}) {
+			callback(id)
+		},
+		OnStoppedLeading: func() {
+			leader, _, err := getCurrentLeader(electionId, namespace, c)
+			if err != nil {
+				glog.Errorf("failed to get leader: %v", err)
+				return
+			}
+			callback(leader)
```
Comment (mikedanese) [inline election/lib/election.go]: if we are a leader and the network partitions, we want to stop leading but we may not be able to observe a new leader if we can't talk to the apiserver. What do we want to report? callback("")? We shouldn't just keep leading.

File: election/example/main.go (hunk 1 above)
Comment (brendandburns) [inline election/example/main.go]: done.

File: election/lib/election.go (hunk 2 above)
Comment (brendandburns) [inline election/lib/election.go]: done.

**Comment by brendandburns** [comment]
Comments addressed, please re-check.

Thanks!
--brendan

**Comment by mikedanese** [comment]
LGTM, I might try to use this to write an e2e test for the leader election client.

**Comment by mikedanese** [comment]
Looks like you need to upgrade the godeps for k8s

**Comment by mikedanese** [comment]
Also do you mean to check in the binary?

**Comment by brendandburns** [comment]
@mikedanese this has been rebased on top of the godeps update elsewhere.  ptal.

thanks!
--brendan

**Comment by mikedanese** [comment]
Some boilerplate failures.

**Comment by brendandburns** [comment]
fixed (I hope)

ptal
thanks!

**Comment by mikedanese** [comment]
LGTM

**Comment by imesh** [comment]
@brendandburns Do we have any documentation on how the leader election works? Is it using etcd internally? Thanks!

**Comment by aledbf** [comment]
@imesh please check this comment https://github.com/kubernetes/kubernetes/blob/master/pkg/client/leaderelection/leaderelection.go#L17

**Comment by imesh** [comment]
Thanks @aledbf!

**Comment by pdeva** [comment]
> @imesh please check this comment https://github.com/kubernetes/kubernetes/blob/master/pkg/client/leaderelection/leaderelection.go#L17

@aledbf the link is broken 😔

**Comment by samchenraising** [comment]
> > @imesh please check this comment https://github.com/kubernetes/kubernetes/blob/master/pkg/client/leaderelection/leaderelection.go#L17
> 
> @aledbf the link is broken 😔

@pdeva, you can browse to the repo from [that day in time](https://github.com/kubernetes/kubernetes/blob/e7d01097dcd2b23c571006792fb2ed8cb85553ef/pkg/client/leaderelection/leaderelection.go#L17)

**Pull request merged**